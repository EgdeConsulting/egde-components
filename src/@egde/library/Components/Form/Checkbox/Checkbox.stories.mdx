import { Meta, Story, Canvas } from '@storybook/addon-docs';
import { LibPath, DocsPageHeader, DocsPageBodyComponents } from '@egde/library';
import { Checkbox } from '@egde/components';
import { useState } from 'react';

export const Template = (args) => {
    const [values, setValues] = useState([...args.value]);
    const handleOnChange = (newValue) => {
        setValues(newValue);
    };
    return (
        <div>
            <Checkbox {...args} onChange={handleOnChange} value={values} />
        </div>
    );
};

export const componentName = 'Checkbox';

<Meta
    title={LibPath.Form + componentName}
    component={Checkbox}
    args={{
        direction: 'column',
        options: ['Option One', 'Option Two', 'Option Three'],
        value: [0],
        onChange: () => {},
        invalidText: '',
        isDisabled: false,
        backgroundColor: '#ffffff',
        autoFocus: false,
        label: 'Checkbox',
        labelTextMargin: '10px',
        captionText: '',
        captionTextMargin: '0px',
        margin: '0px',
        padding: '0px',
        width: '100%',
        minWidth: '0px',
        maxWidth: '100%',
    }}
/>

<DocsPageHeader
    title={componentName}
    description="This component is used to create a checklist where multiple answers can be checked off."
    addImportantNote={false}
    variant="stable"
/>

<Canvas withToolbar>
    <Story name="Base">{Template.bind({})}</Story>
</Canvas>

<DocsPageBodyComponents addStories={true} />

### Checkbox Row

<Canvas withToolbar>
    <Story
        name="Checkbox Row"
        args={{
            direction: 'row',
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Invalide Checkbox

<Canvas withToolbar>
    <Story
        name="Invalid Checkbox"
        args={{
            invalidText: 'Invalid',
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Disabled Checkbox

<Canvas withToolbar>
    <Story
        name="Disabled Checkbox"
        args={{
            isDisabled: true,
            value: [],
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Multiple Checked

<Canvas withToolbar>
    <Story
        name="Multiple Checked"
        args={{
            value: [0, 1, 2],
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>
