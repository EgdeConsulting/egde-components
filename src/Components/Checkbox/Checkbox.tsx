import React, { ReactElement } from 'react';
import { Checkbox as ChakraCheckbox, Stack } from '@chakra-ui/react';
import { CheckboxProps } from 'Types';
import { BaseInput } from '../Input/BaseInput';

const Checkbox = (props: CheckboxProps): ReactElement => {
  const {
    direction = 'column',
    options,
    value,
    onChange,
    invalidText,
    isDisabled,
    backgroundColor,
    autoFocus,
  } = props;

  function handleOnChange(optionIndex: number): void {
    let newValue = [];
    if (value.includes(optionIndex)) {
      newValue = value.filter((e) => e !== optionIndex);
    } else {
      newValue = [...value, optionIndex];
    }
    onChange(newValue);
  }

  return (
    <BaseInput {...props}>
      <Stack direction={direction}>
        {options.map((option, index) => {
          return (
            <ChakraCheckbox
              isInvalid={!!invalidText}
              isDisabled={isDisabled}
              key={`chk${index}`}
              onChange={() => handleOnChange(index)}
              isChecked={value.includes(index)}
              backgroundColor={backgroundColor}
              autoFocus={autoFocus}
            >
              {option}
            </ChakraCheckbox>
          );
        })}
      </Stack>
    </BaseInput>
  );
};

export { Checkbox };
