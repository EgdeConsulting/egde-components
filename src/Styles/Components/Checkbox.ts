const Checkbox = {
    baseStyle: {
        control: {
            backgroundColor: 'white',
            borderColor: 'blue',
            borderRadius: 'sm',
            _checked: {
                backgroundColor: 'blue',
                borderColor: 'blue',
                _disabled: {
                    backgroundColor: 'lightblue',
                    borderColor: 'lightblue',
                },
                _hover: {
                    backgroundColor: 'darkblue',
                    borderColor: 'darkblue',
                },
                _invalid: {
                    backgroundColor: 'white',
                    borderColor: 'red',
                    borderWidth: '2px',
                },
            },
            _disabled: {
                backgroundColor: 'lightblue',
                borderColor: 'lightblue',
                _invalid: {
                    backgroundColor: 'lightblue',
                    borderColor: 'lightblue',
                },
            },
            _hover: {
                borderColor: 'darkblue',
                _checked: {
                    backgroundColor: 'darkblue',
                },
                _disabled: {
                    backgroundColor: 'lightblue',
                    borderColor: 'lightblue',
                },
                _invalid: {
                    borderWidth: '2px',
                    borderColor: 'red',
                },
            },
            _invalid: {
                borderWidth: '2px',
                borderColor: 'red',
            },
        },
    },
};

export { Checkbox };
