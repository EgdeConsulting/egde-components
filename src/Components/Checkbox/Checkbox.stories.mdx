import {
    Meta,
    Story,
    Props,
    Canvas,
    Preview,
} from '@storybook/addon-docs/blocks';
import { Checkbox } from './Checkbox';
import { ComponentStability } from 'Icons';
import { LibPath } from 'Types';
import { useState } from 'react';

export const Template = (args) => {
    const [values, setValues] = useState([...args.value]);
    const handleOnChange = (newValue) => {
        setValues(newValue);
    };
    return (
        <div>
            <Checkbox {...args} onChange={handleOnChange} value={values} />
        </div>
    );
};

<Meta
    title={LibPath.Form + 'Checkbox'}
    component={Checkbox}
    args={{
        direction: 'column',
        options: [
            'Checkbox Option One',
            'Checkbox Option Two',
            'Checkbox Option Three',
        ],
        value: [0],
        onChange: () => {},
        invalidText: '',
        isDisabled: false,
        backgroundColor: '#ffffff',
        autoFocus: false,
        label: 'Checkbox',
        labelTextMargin: '0px',
        captionText: '',
        captionTextMargin: '0px',
        margin: '0px',
        padding: '0px',
        width: '100%',
        minWidth: '0px',
        maxWidth: '100%',
    }}
/>

# Checkbox

<ComponentStability variant="stable" />

<hr />

This component is used to create a checklist where multiple answers can be checked off.

<Preview withToolbar>
    <Story name="Base">{Template.bind({})}</Story>
</Preview>

Change some of the props and see how the `Base` component changes.

<Props story="Base" />

# Stories

<hr />

Below are shown different implementations of the component in regards to varying
use cases.

<br />

### Checkbox Row

<Canvas withToolbar>
    <Story
        name="Checkbox Row"
        args={{
            direction: 'row',
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Invalide Checkbox

<Canvas withToolbar>
    <Story
        name="Invalid Checkbox"
        args={{
            invalidText: 'Invalide',
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Disabled Checkbox

<Canvas withToolbar>
    <Story
        name="Disabled Checkbox"
        args={{
            isDisabled: true,
            value: [],
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Multiple Checked

<Canvas withToolbar>
    <Story
        name="Multiple Checked"
        args={{
            value: [0, 1, 2],
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>
