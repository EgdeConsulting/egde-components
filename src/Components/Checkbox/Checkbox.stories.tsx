import React from 'react';
import { LibPath } from 'Types';
import { Checkbox } from './Checkbox';
import { ComponentStory, ComponentMeta } from '@storybook/react';

export default {
  title: LibPath.Input || 'undefined',
  component: Checkbox,
  argTypes: {
    direction: { defaultValue: 'column' },
    options: {
      defaultValue: [
        'Checkbox Option One',
        'Checkbox Option Two',
        'Checkbox Option Three',
      ],
    },
    value: { defaultValue: '' },
    onChange: { defaultValue: () => {} },
    invalidText: { defaultValue: '' },
    isDisabled: { defaultValue: false },
    backgroundColor: { defaultValue: '#fff' },
    autoFocus: { defaultValue: false },
  },
} as ComponentMeta<typeof Checkbox>;

const Template: ComponentStory<typeof Checkbox> = (args) => (
  <Checkbox {...args} />
);

export const CheckboxStd = Template.bind({});
CheckboxStd.args = {};
