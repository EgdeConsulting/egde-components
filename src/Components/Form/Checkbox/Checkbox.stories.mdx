import { Meta, Story, Canvas } from '@storybook/addon-docs';
import { DocsPageHeader, DocsPageBodyComponents } from 'Docs';
import { FormComponentDemo } from './FormComponentDemo';
import { LibPath } from 'Types';
import { Checkbox } from 'Components';
import { useState } from 'react';
import action from '@storybook/addon-actions';

export const Template = (args) => <Checkbox {...args} />;

export const componentName = 'Checkbox';

<Meta
    title={LibPath.Form + componentName}
    component={Checkbox}
    args={{
        direction: 'column',
        options: ['Option One', 'Option Two', 'Option Three'],
        value: [0],
        invalidText: '',
        isDisabled: false,
        backgroundColor: '#ffffff',
        autoFocus: false,
        label: 'Checkbox',
        labelTextMargin: '10px',
        captionText: '',
        captionTextMargin: '0px',
        margin: '0px',
        padding: '0px',
        width: '100%',
        minWidth: '0px',
        maxWidth: '100%',
    }}
    argtypes={{
        onChange: {
            table: {
                category: 'Actions',
            },
        },
    }}
    decorators={[
        (Story, { args }) => (
            <FormComponentDemo
                initialValue={args.values}
                property="values"
                type="checkbox"
            >
                {Story()}
            </FormComponentDemo>
        ),
    ]}
/>

<DocsPageHeader
    title={componentName}
    description="This component is used to create a checklist where multiple answers can be checked off."
    addImportantNote={false}
    variant="stable"
/>

<Canvas withToolbar>
    <Story name="Base">{Template.bind({})}</Story>
</Canvas>

<DocsPageBodyComponents addStories={true} />

### Checkbox Row

<Canvas withToolbar>
    <Story
        name="Checkbox Row"
        args={{
            direction: 'row',
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Invalide Checkbox

<Canvas withToolbar>
    <Story
        name="Invalid Checkbox"
        args={{
            invalidText: 'Invalid',
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Disabled Checkbox

<Canvas withToolbar>
    <Story
        name="Disabled Checkbox"
        args={{
            isDisabled: true,
            value: [],
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Multiple Checked

<Canvas withToolbar>
    <Story
        name="Multiple Checked"
        args={{
            value: [0, 1, 2],
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>
