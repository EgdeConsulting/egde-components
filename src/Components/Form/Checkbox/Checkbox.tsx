import React, { ReactElement } from 'react';
import { Checkbox as ChakraCheckbox, FormLabel, Stack } from '@chakra-ui/react';
import { CheckboxProps } from 'SharedTypes';
import { BaseInput } from 'Components';

export function Checkbox(props: CheckboxProps): ReactElement {
    const {
        label,
        labelTextMargin,
        direction = 'column',
        options,
        value,
        onChange,
        invalidText,
        isDisabled,
        captionText,
        margin,
        padding,
        width,
        minWidth,
        backgroundColor,
        autoFocus,
    } = props;

    function handleOnChange(optionIndex: number): void {
        let newValue = [];
        if (value.includes(optionIndex)) {
            newValue = value.filter((e) => e !== optionIndex);
        } else {
            newValue = [...value, optionIndex];
        }
        onChange(newValue);
    }

    return (
        <BaseInput
            invalidText={invalidText}
            captionText={captionText}
            margin={margin}
            padding={padding}
            width={width}
            minWidth={minWidth}
            isGroup
        >
            <FormLabel
                fontWeight='bold'
                color='darkgrey'
                margin={labelTextMargin || '0 0 8px 0'}
                as='legend'
            >
                {label}
            </FormLabel>
            <Stack direction={direction}>
                {options.map((option, index) => {
                    return (
                        <ChakraCheckbox
                            isInvalid={!!invalidText}
                            isDisabled={isDisabled}
                            key={`chk${index}`}
                            onChange={() => handleOnChange(index)}
                            isChecked={value.includes(index)}
                            backgroundColor={backgroundColor}
                            autoFocus={autoFocus}
                        >
                            {option}
                        </ChakraCheckbox>
                    );
                })}
            </Stack>
        </BaseInput>
    );
}
