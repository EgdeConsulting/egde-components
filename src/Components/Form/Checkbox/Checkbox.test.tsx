import React from 'react';
import { render } from '@testing-library/react';
import { Checkbox } from './Checkbox';


// Snapshot Test
describe('Snapshot Checkbox', () => {
  it('should render correctly', () => {
    const tree = render(<Checkbox 
        direction= 'column'
        options= {['Checkbox Option One', 'Checkbox Option Two', 'Checkbox Option Three']}
        value= {[0]}
        onChange= {() => {}}
        invalidText= ''
        isDisabled= {false}
        backgroundColor= '#ffffff'
        autoFocus={false}
        label= 'Checkbox'
        labelTextMargin= '0px'
        captionText= ''
        captionTextMargin= '0px'
        margin= '0px'
        padding= '0px'
        width= '100%'
        minWidth= '0px'
        maxWidth= '100%'
        />);
    expect(tree).toMatchSnapshot();
  });
});
