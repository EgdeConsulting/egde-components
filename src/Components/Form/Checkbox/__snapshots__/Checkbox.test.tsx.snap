// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component test should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <div
        class="css-k456he"
      >
        <p
          class="chakra-text css-154u26z"
        >
          Checkbox Tester
        </p>
        <div
          class="css-1o2srd0"
        >
          <div
            class="chakra-stack css-n21gh5"
          >
            <label
              class="chakra-checkbox css-re7v4p"
            >
              <input
                aria-disabled="false"
                aria-invalid="false"
                class="chakra-checkbox__input"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-checkbox__control css-dnty2r"
              />
              <span
                class="chakra-checkbox__label css-1oeb2oe"
              >
                Checkbox Option One
              </span>
            </label>
            <label
              class="chakra-checkbox css-re7v4p"
            >
              <input
                aria-disabled="false"
                aria-invalid="false"
                class="chakra-checkbox__input"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-checkbox__control css-dnty2r"
              />
              <span
                class="chakra-checkbox__label css-1oeb2oe"
              >
                Checkbox Option Two
              </span>
            </label>
          </div>
        </div>
        <p
          class="chakra-text css-1vrfh4l"
        >
          Caption text
        </p>
        
      </div>
      <button
        type="button"
      >
        Toggle disabled
      </button>
      <button
        type="button"
      >
        Invalidate
      </button>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-k456he"
    >
      <p
        class="chakra-text css-154u26z"
      >
        Checkbox Tester
      </p>
      <div
        class="css-1o2srd0"
      >
        <div
          class="chakra-stack css-n21gh5"
        >
          <label
            class="chakra-checkbox css-re7v4p"
          >
            <input
              aria-disabled="false"
              aria-invalid="false"
              class="chakra-checkbox__input"
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="checkbox"
              value=""
            />
            <span
              aria-hidden="true"
              class="chakra-checkbox__control css-dnty2r"
            />
            <span
              class="chakra-checkbox__label css-1oeb2oe"
            >
              Checkbox Option One
            </span>
          </label>
          <label
            class="chakra-checkbox css-re7v4p"
          >
            <input
              aria-disabled="false"
              aria-invalid="false"
              class="chakra-checkbox__input"
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="checkbox"
              value=""
            />
            <span
              aria-hidden="true"
              class="chakra-checkbox__control css-dnty2r"
            />
            <span
              class="chakra-checkbox__label css-1oeb2oe"
            >
              Checkbox Option Two
            </span>
          </label>
        </div>
      </div>
      <p
        class="chakra-text css-1vrfh4l"
      >
        Caption text
      </p>
      
    </div>
    <button
      type="button"
    >
      Toggle disabled
    </button>
    <button
      type="button"
    >
      Invalidate
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
